ibmcloud: build se-image with docker and packer

liudalibj opened the pull request:
- update ubuntu packer hcl
- add build-se-image script
- update related documents
- replace `CAA_SRC_REF` to `CAA_SRC_BRANCH` to macth document

Fixes: #877

 Signed-off-by: Da Li Liu <liudali@cn.ibm.com>

yoheiueda reviewed (commented):
Thank you very much for preparing this great enhancement.

At first glance, your approach looks as follows.

1. Initiates creation of a non-SE image with packer
2. Invokes `build-se-image.sh` in a mounted filesystem image which is being created by packer
3. Copies the content of the non-SE image with the `dd` command to another encrypted image

Is is possible to copy the content from a non-SE image to an encrypted image while the source image is not mounted?

I don't think it is safe to copy a block device to another device with `dd` while it is mounted with a read-write mode.

The steps will be like this.

1. Create a non-SE image with packer
2. Use qemu-nbd to attach the image to a nbd device
3. Copy the content of the image to an encrypte image with `dd`.

liudalibj commented:
> Is is possible to copy the content from a non-SE image to an encrypted image while the source image is not mounted?
> The steps will be like this.
> 
> Create a non-SE image with packer
> Use qemu-nbd to attach the image to a nbd device
> Copy the content of the image to an encrypte image with dd.

This is the first thing I tried, but I failed.
What I tried:
- Build `podvm_builder` image
- Start one container with the `podvm_builder`
- Create a non-SE image with packer inside the container
- Try to use `qemu-nbd`, but I was blocked at `modprobe nbd`
It said:
```
root@19078d46dc8d:/src# modprobe nbd
modprobe: FATAL: Module nbd not found in directory /lib/modules/5.4.0-1044-ibm
```

To address the nbd issue, I tried second way:
- Build `podvm_builder` image
- Start one container with the `podvm_builder`
- Run packer command inside the container and build the SE image inside the qemu s390x instance by using the script same as origin SE build script
- Copy the built SE-image from the qemu s390x instance to the container
- Clean and create non-SE image with packer.

This way will take long time, the base image will be downloaded again inside the qemu s390x instance, the `copy-files.sh` & `misc-settings.sh` will be run twice(one from packer provision , second by the se-build script)

The third way just as this PR.

liudalibj commented on podvm/qcow2/build-se-image.sh:
```diff
@@ -0,0 +1,165 @@
+
+# TOOTOTO
+echo "Copying root partition with dd"
+sudo dd if="${src_nbd}" of="/dev/mapper/${LUKS_NAME}" bs=$((1024*1024))
```
I tied use the `tar` command to copy the running root filesystem to another mount device, but I meet follow errors:
```
peerpod@ubuntu:~$ sudo tar -cf - "${tar_opts[@]}" --sort=none -C "$src_mnt" . | tar -xf - "${tar_opts[@]}" --preserve-order  -C "$dst_mnt"
tar: ./lib: Cannot create symlink to ‘usr/lib’: Permission denied
tar: ./boot: Cannot mkdir: Permission denied
tar: ./boot: Cannot mkdir: Permission denied
tar: ./boot/System.map-5.4.0-147-generic: Cannot open: No such file or directory
tar: ./boot: Cannot mkdir: Permission denied
...
```

yoheiueda replied:
Copying a system filesystem with `tar` needs a lot of options. I think we need at least [`--one-file-system`](https://serverfault.com/a/561735) in addition to the options I used below.

https://github.com/confidential-containers/cloud-api-adaptor/blob/e7183af5889436de4b750d6fb0c79fe44c00ec97/ibmcloud/image/build.sh#L287-L289

I think `--one-file-system` is necessary, since various additional filesystems such as `/tmp` and `/proc` are mounted in a running Linux operating system.

You also need use `sudo` to run the second `tar`.

yoheiueda replied:
I think the `tar` approach is safer than the `fsfreeze` approach, if it is functional.

liudalibj replied:
Thanks @yoheiueda ,
I updated the codes to use `tar` approach with `sudo mount --bind -o ro / ${src_mnt}`, the tar command output:
```
tar: ./var/snap/lxd/common/lxd/unix.socket: socket ignored
```
The built out se-image seems bigger than the `dd` approach, it's 3.5G
```
-rw-r--r-- 1 root root 3.5G May  8 06:27 se-podvm-2413ed0-dirty-s390x.qcow2
```

The built out qcow2 image works as expected, e2e test are passed with it.

yoheiueda replied:
@liudalibj Thank you for the update!

I thought the se-image would be 100GB, since luksFormat leads to allocate all disk space of the QCOW2 image. I didn't think QCOW2's copy-on-write mechanism works for encrypted disks. Do you have any idea how it works?

Supposing lazy allocation works for encrypted disks, the difference between the two approaches may be the actual partition sizes. I think the source partition of the `dd` command was not 100GB. Right?

https://github.com/confidential-containers/cloud-api-adaptor/blob/bc64906141e15de4388ef53a281a96ed02b99f44/podvm/qcow2/ubuntu/variables.pkr.hcl#L16-L19

I think 100GB is a limitation for IBM Cloud VPC, and we don't need full 100GB space for peer pod VM components. 
A possible work around is to create a 6GB partition, and encrypted it with LUKS. We can keep the remaining spare unused.

liudalibj replied:
@yoheiueda I don't know how it works that why with luksFormat we didn't get one 100G qcow2 file.

And I tried your suggestion, to make the root partition to 6G and then encrypted it.
The final output qcow2 file is reduced to 1.7 G, thanks!
```
-rw-r--r-- 1 root root 1.7G May 10 06:21 se-podvm-0b154ef-dirty-s390x.qcow2
```
So we have one unused `data` partition after create one SE VSI with the built out image.
```
peerpod@se-test-38:~$ lsblk
NAME                                          MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINT
loop0                                           7:0    0 54.6M  1 loop  /snap/core20/1881
loop1                                           7:1    0 77.8M  1 loop  /snap/lxd/24063
loop2                                           7:2    0 49.6M  1 loop  /snap/snapd/19126
vda                                           252:0    0  100G  0 disk
├─vda1                                        252:1    0  255M  0 part  /boot-se
├─vda2                                        252:2    0    6G  0 part
│ └─luks-aff936d5-0d05-4ea3-b038-192ba4df6f41 253:0    0    6G  0 crypt /
└─vda3                                        252:3    0 86.9G  0 part
vdb                                           252:16   0  366K  0 disk
vdc                                           252:32   0   44K  0 disk
peerpod@se-test-38:~$ exit
```

liudalibj commented on podvm/qcow2/build-se-image.sh:
```diff
@@ -0,0 +1,165 @@
+export tmp_nbd="/dev/$device"
+export dst_mnt=$workdir/dst_mnt
+# /dev/vda1 or /dev/vdb1
+export src_nbd=$(mount |grep "on / type ext4 (rw,relatime)" | awk '{ print $1 }')
```
Is there a good way that we copy one running root filesystem `/` to another mount device?

yoheiueda replied:
Could you try to remount the root filesystem with read-only mode here like this?

```
mount -o remount,ro /
```
If it is possible, copying with `dd` becomes much safer.

yoheiueda replied:
Yet another option is `fsfreeze`.
https://man7.org/linux/man-pages/man8/fsfreeze.8.html

liudalibj replied:
Failed to remount.
```
peerpod@ubuntu:~$ sudo mount -o remount,ro /
mount: /: mount point is busy.
```

`fsfreeze` is worked well.

yoheiueda replied:
Did you see any error messages like below in dmesg or syslog when you boot with a image built with `fsfreeze`? 

```
/dev/??? was not cleanly unmounted, check forced
```

liudalibj replied:
This is SE-image, I can't ssh to the created ZVSI with it.

bpradipt commented:
> @bpradipt do you have any comments? I am wondering why qcow2 image is built in `docker build`, instead of `docker run` in the current build mechanism.

Hmm I don't think we ever thought of using `docker run` primarily because we were using build args to modify the build process.

bpradipt commented:
@liudalibj is there a document that I can read to understand what is required to build an se-image ? or if you can describe the highlevel steps? I might be able to provide practical suggestions.

huoqifeng commented:
> @liudalibj is there a document that I can read to understand what is required to build an se-image ? or if you can describe the highlevel steps? I might be able to provide practical suggestions.

@bpradipt I think this is the document https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/ibmcloud/SECURE_EXECUTION.md

liudalibj commented:
> > @liudalibj is there a document that I can read to understand what is required to build an se-image ? or if you can describe the highlevel steps? I might be able to provide practical suggestions.
> 
> @bpradipt I think this is the document https://github.com/confidential-containers/cloud-api-adaptor/blob/staging/ibmcloud/SECURE_EXECUTION.md

Thanks @huoqifeng the update document is in this pr, the link: https://github.com/confidential-containers/cloud-api-adaptor/blob/b159ec45bd0f8132b32cb3ec44ea210e843b85c1/ibmcloud/SECURE_EXECUTION.md

liudalibj commented:
To verify this pr, full steps can follow https://github.com/liudalibj/cloud-api-adaptor/blob/se-image-packer/ibmcloud/SECURE_EXECUTION.md
Simple steps:
1. 
```
docker build --no-cache -t podvm_builder -f  ./Dockerfile.podvm_builder .
```
2.
```
docker build -t podvm_binaries_s390x \
         --build-arg BUILDER_IMG=podvm_builder \
         --build-arg ARCH=s390x \
         -f Dockerfile.podvm_binaries .
```
3. Download host document refer `Download host key document from Resource Link` [at here](https://github.com/liudalibj/cloud-api-adaptor/blob/se-image-packer/ibmcloud/SECURE_EXECUTION.md)
4. put host document to `podvm/files`
5. 
```
docker build -t se_podvm_s390x \
         --build-arg ARCH=s390x \
         --build-arg CAA_SRC=https://github.com/liudalibj/cloud-api-adaptor \
         --build-arg CAA_SRC_BRANCH=se-image-packer \
         --build-arg SE_BOOT=1 \
         --build-arg BUILDER_IMG=podvm_builder \
         --build-arg BINARIES_IMG=podvm_binaries_s390x \
         --build-arg UBUNTU_IMAGE_URL="" \
         --build-arg UBUNTU_IMAGE_CHECKSUM="" \
         -f Dockerfile.podvm .
```
the output looks like:
```
...
Removing intermediate container 2c0ed25d2250
 ---> 12d8c6dbe7e7
Step 25/28 : COPY .dummy files/*.crt /src/cloud-api-adaptor/podvm/files/
 ---> 6bd0e9e99b3a
Step 26/28 : RUN cd cloud-api-adaptor/podvm &&      make BINARIES= PAUSE_BUNDLE= image
 ---> Running in fa28447bd11a
Makefile.inc:28: A foreign ARCH was passed, but no CC alternative. Using s390x-linux-gnu-gcc as best guess
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  5394  100  5394    0     0   250k      0 --:--:-- --:--:-- --:--:--  250k
defined
rm -fr output
rm -f cloud-init.img
cloud-localds cloud-init.img qcow2/userdata.cfg
mkdir -p toupload
qemu-img create -f qcow2 "se-podvm-0b154ef-dirty-s390x.qcow2" 100G
Formatting 'se-podvm-0b154ef-dirty-s390x.qcow2', fmt=qcow2 size=107374182400 cluster_size=65536 lazy_refcounts=off refcount_bits=16
packer build -var boot_wait=600s  -var qemu_image_name=podvm-0b154ef-dirty-s390x.qcow2 -var cloud_image_url=https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-s390x.img -var cloud_image_checksum=378f0b1dd97dc11f6828f5d7e6b7a97d9f31ddaaf8481ea9ff68d9e999ec62ea -var se_boot=1 -var disk_size=2560  -var machine_type=s390-ccw-virtio -var qemu_binary=qemu-system-s390x qcow2/ubuntu
qemu.ubuntu: output will be in this color.

==> qemu.ubuntu: Retrieving ISO
==> qemu.ubuntu: Trying https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-s390x.img
==> qemu.ubuntu: Trying https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-s390x.img?checksum=sha256%3A378f0b1dd97dc11f6828f5d7e6b7a97d9f31ddaaf8481ea9ff68d9e999ec62ea
==> qemu.ubuntu: https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-s390x.img?checksum=sha256%3A378f0b1dd97dc11f6828f5d7e6b7a97d9f31ddaaf8481ea9ff68d9e999ec62ea => /root/.cache/packer/a9625202848d3c63b0bdd4b4b2c8f438d311bb60.iso
==> qemu.ubuntu: Copying hard drive...
==> qemu.ubuntu: Resizing hard drive...
==> qemu.ubuntu: Found port for communicator (SSH, WinRM, etc): 2706.
==> qemu.ubuntu: Looking for available port between 5900 and 6000 on 127.0.0.1
==> qemu.ubuntu: Starting VM, booting disk image
    qemu.ubuntu: The VM will be run headless, without a GUI. If you want to
    qemu.ubuntu: view the screen of the VM, connect via VNC without a password to
    qemu.ubuntu: vnc://127.0.0.1:5965
==> qemu.ubuntu: Overriding default Qemu arguments with qemuargs template option...
==> qemu.ubuntu: Waiting 10m0s for boot...
==> qemu.ubuntu: Connecting to VM via VNC (127.0.0.1:5965)
==> qemu.ubuntu: Typing the boot commands over VNC...
    qemu.ubuntu: Not using a NetBridge -- skipping StepWaitGuestAddress
==> qemu.ubuntu: Using SSH communicator to connect: 127.0.0.1
==> qemu.ubuntu: Waiting for SSH to become available...
==> qemu.ubuntu: Connected to SSH!
==> qemu.ubuntu: Running local shell script: /tmp/packer-shell1650844487
==> qemu.ubuntu: Uploading ./toupload => /tmp/
==> qemu.ubuntu: Provisioning with shell script: /tmp/packer-shell1524450936
==> qemu.ubuntu: Uploading qcow2/copy-files.sh => ~/copy-files.sh
==> qemu.ubuntu: Provisioning with shell script: /tmp/packer-shell3814779015
==> qemu.ubuntu: Uploading qcow2/misc-settings.sh => ~/misc-settings.sh
==> qemu.ubuntu: Provisioning with shell script: /tmp/packer-shell3912101311
    qemu.ubuntu: passwd: password expiry information changed.
==> qemu.ubuntu: Removed /etc/systemd/system/timers.target.wants/apt-daily.timer.
==> qemu.ubuntu: Removed /etc/systemd/system/timers.target.wants/apt-daily-upgrade.timer.
==> qemu.ubuntu: Removed /etc/systemd/system/multi-user.target.wants/snapd.service.
==> qemu.ubuntu: Removed /etc/systemd/system/cloud-final.service.wants/snapd.seeded.service.
==> qemu.ubuntu: Removed /etc/systemd/system/multi-user.target.wants/snapd.seeded.service.
==> qemu.ubuntu: Removed /etc/systemd/system/multi-user.target.wants/snap.lxd.activate.service.
==> qemu.ubuntu: Uploading qcow2/build-se-image.sh => ~/build-se-image.sh
==> qemu.ubuntu: Provisioning with shell script: /tmp/packer-shell3903816458
    qemu.ubuntu: Building SE podvm image for s390x
    qemu.ubuntu: Finding host key files
    qemu.ubuntu: found host key file: "/tmp/files/HKD-8562-024B878-eu-gb-2.crt"
    qemu.ubuntu: Creating boot-se and root partitions
==> qemu.ubuntu: Information: You may need to update /etc/fstab.
==> qemu.ubuntu:
    qemu.ubuntu: Waiting for the two partitions to show up
    qemu.ubuntu: Formatting boot-se partition
==> qemu.ubuntu: mke2fs 1.45.5 (07-Jan-2020)
    qemu.ubuntu: Discarding device blocks: done
    qemu.ubuntu: Creating filesystem with 65280 4k blocks and 65280 inodes
    qemu.ubuntu: Filesystem UUID: 2cf4624d-f8a2-4cc3-bd59-6a8e2e93da6a
    qemu.ubuntu: Superblock backups stored on blocks:
    qemu.ubuntu: 	32768
    qemu.ubuntu:
    qemu.ubuntu: Allocating group tables: done
    qemu.ubuntu: Writing inode tables: done
    qemu.ubuntu: Creating journal (4096 blocks): done
    qemu.ubuntu: Writing superblocks and filesystem accounting information: done
    qemu.ubuntu:
    qemu.ubuntu: Setting up encrypted root partition
    qemu.ubuntu: Setting luks name for root partition
    qemu.ubuntu: luks name is: luks-1dd0d730-6cf2-4553-bffc-97ab1e42b3de
    qemu.ubuntu: Copying the root filesystem
==> qemu.ubuntu: mke2fs 1.45.5 (07-Jan-2020)
    qemu.ubuntu: Creating filesystem with 24344320 4k blocks and 6086656 inodes
    qemu.ubuntu: Filesystem UUID: ed855ab5-85c3-4633-b517-cd086ed5b508
    qemu.ubuntu: Superblock backups stored on blocks:
    qemu.ubuntu: 	32768, 98304, 163840, 229376, 294912, 819200, 884736, 1605632, 2654208,
    qemu.ubuntu: 	4096000, 7962624, 11239424, 20480000, 23887872
    qemu.ubuntu:
    qemu.ubuntu: Allocating group tables: done
    qemu.ubuntu: Writing inode tables: done
    qemu.ubuntu: Creating journal (131072 blocks): done
    qemu.ubuntu: Writing superblocks and filesystem accounting information: done
    qemu.ubuntu:
==> qemu.ubuntu: tar: ./var/snap/lxd/common/lxd/unix.socket: socket ignored
    qemu.ubuntu: Partition copy complete
    qemu.ubuntu: Preparing secure execution boot image
    qemu.ubuntu: Adding fstab
    qemu.ubuntu: Adding luks keyfile for fs
    qemu.ubuntu: Prep files needed for mkinitrd
    qemu.ubuntu: Updating initial ram disk
    qemu.ubuntu: update-initramfs: Generating /boot/initrd.img-5.4.0-148-generic
==> qemu.ubuntu: Error: Image file '/boot-se/se.img' in section 'linux': No such file or directory
==> qemu.ubuntu: Using config file '/etc/zipl.conf'
==> qemu.ubuntu: run-parts: /etc/initramfs/post-update.d//zz-zipl exited with return code 1
==> qemu.ubuntu: !!! Bootloader install errors prior to this line are intentional !!!!!
    qemu.ubuntu: Generating an IBM Secure Execution image
    qemu.ubuntu: Create SE boot image
==> qemu.ubuntu: WARNING: host-key document verification is disabled. Your workload is not secured.
    qemu.ubuntu: Running zipl to prepare boot partition
    qemu.ubuntu: Building bootmap in '/boot-se'
    qemu.ubuntu: Adding IPL section
    qemu.ubuntu:   kernel image......: /boot-se/se.img
    qemu.ubuntu:   component address:
    qemu.ubuntu:     heap area.......: 0x00002000-0x00005fff
    qemu.ubuntu:     stack area......: 0x0000f000-0x0000ffff
    qemu.ubuntu:     internal loader.: 0x0000a000-0x0000dfff
    qemu.ubuntu:     parameters......: 0x00009000-0x000091ff
    qemu.ubuntu:     kernel image....: 0x00010000-0x01eb4fff
    qemu.ubuntu: Preparing boot device: vda.
    qemu.ubuntu: Done.
    qemu.ubuntu: Clean luks keyfile
    qemu.ubuntu: Closing encrypted root partition
==> qemu.ubuntu: Gracefully halting virtual machine...
==> qemu.ubuntu: Converting hard drive...
==> qemu.ubuntu: Running post-processor: post-build-se-image (type shell-local)
==> qemu.ubuntu (shell-local): Running local shell script: qcow2/build-se-image-post.sh
    qemu.ubuntu (shell-local): SE podvm image is built: output/se-podvm-0b154ef-dirty-s390x.qcow2
Build 'qemu.ubuntu' finished after 25 minutes 51 seconds.

==> Wait completed after 25 minutes 51 seconds

==> Builds finished. The artifacts of successful builds are:
--> qemu.ubuntu: VM files in directory: output
--> qemu.ubuntu: VM files in directory: output
rm -fr toupload
rm -f cloud-init.img
Removing intermediate container fa28447bd11a
 ---> 2e769f7a9436
Step 27/28 : FROM scratch
 --->
Step 28/28 : COPY --from=podvm_builder /src/cloud-api-adaptor/podvm/output/*.qcow2 /
 ---> e97ee9afa3d7
Successfully built e97ee9afa3d7
Successfully tagged se_podvm_s390x:latest
root@liudali-e2e-image:~/cloud-api-adaptor/podvm#
```

bpradipt commented on podvm/Dockerfile.podvm:
```diff
@@ -13,27 +13,29 @@ FROM ${BINARIES_IMG} AS podvm_binaries
 ARG ARCH=x86_64
 ARG CAA_SRC=""
-ARG CAA_SRC_REF=""
+ARG CAA_SRC_BRANCH=""
```
You can use either branch name or commit id for `CAA_SRC_REF`

liudalibj replied:
Got it, thanks @bpradipt  so that I will keep using `CAA_SRC_REF ` and then update related Document.

bpradipt commented on podvm/Makefile:
```diff
@@ -80,12 +87,12 @@ ifdef UEFI_FIRMWARE_LOCATION
 	rm -f cloud-init.img
 	cloud-localds cloud-init.img qcow2/userdata.cfg
 	mkdir -p toupload
+	qemu-img create -f qcow2 "se-${IMAGE_FILE}" 100G
```
So an empty base image will be needed for SE ?
Is it possible to put it under a conditional?

liudalibj replied:
Yes this se-${IMAGE_FILE} is needed for SE. 
And as we attach this empty qcow2 `se-${IMAGE_FILE}` file as a block device to the qemu s390x instance, so we need create if before packer.

liudalibj replied:
I am trying to use a condition for SE image build.

liudalibj replied:
done

bpradipt commented on podvm/qcow2/build-se-image.sh:
```diff
@@ -0,0 +1,162 @@
+if [ "${SE_BOOT:-0}" != "1" ]; then
+    exit 0
+elif [ "${ARCH}" != "s390x" ]; then
+    echo "Build SE podvm image for $ARCH is not support yet"
```
May be a better message is "Building of SE podvm image is only supported for s390x" ?

liudalibj replied:
done

yoheiueda reviewed (commented):
Thank you very much for the instructs for testing the generated SE image.

I followed the instructions to build an SE image on x86 machine, and used it with a Kubernetes cluster on jp-tok-2. I successfully deployed a SE peer pod VM to run the nginx demo.

Overall, It looks good to me. I have some minor comments.

liudalibj commented on podvm/qcow2/ubuntu/qemu-ubuntu.pkr.hcl:
```diff
@@ -1,50 +1,52 @@
+      ["-device", "virtio-blk,drive=virtio-drive,id=virtio-disk0,bootindex=1"],
+      ["-device", "virtio-blk,drive=se-virtio-drive,id=virtio-disk1"],
+      ["-drive", "file=${var.output_directory}/${var.qemu_image_name},if=none,cache=writeback,discard=ignore,format=qcow2,id=virtio-drive"],
+      ["-drive", "file=se-${var.qemu_image_name},if=none,cache=writeback,discard=ignore,format=qcow2,id=se-virtio-drive"],
```
I am trying to use a condition here.

liudalibj replied:
done

liudalibj commented on podvm/qcow2/ubuntu/qemu-ubuntu.pkr.hcl:
```diff
@@ -86,12 +88,40 @@ build {
     ]
   }
 
+  provisioner "file" {
+    source      = "qcow2/build-se-image.sh"
+    destination = "~/build-se-image.sh"
+  }
+
+  provisioner "shell" {
+    remote_folder = "~"
+    environment_vars = [
+      "SE_BOOT=${var.se_boot}",
+      "ARCH=${var.target_arch}"
+    ]
+    inline = [
+      "sudo -E bash ~/build-se-image.sh"
+    ]
+  }
+
+  post-processors {
+    post-processor "shell-local" {
+      name = "post-build-se-image"
+      script = "qcow2/build-se-image-post.sh"
+      environment_vars = [
+        "SE_BOOT=${var.se_boot}",
+        "ARCH=${var.target_arch}",
+        "OUTPUT_DIRECTORY=${var.output_directory}",
+        "IMAGE_NAME=${var.qemu_image_name}"
+      ]
+    }
+  }
```
I am trying to use a condition here.

liudalibj replied:
Didn't found good way to only run the post-process for se_boot.
Found one similar issue https://github.com/hashicorp/packer/issues/10583
and some references:
https://developer.hashicorp.com/packer/guides/packer-on-cicd/pipelineing-builds#chaining-together-several-of-the-same-builders-to-make-save-points
https://developer.hashicorp.com/packer/docs/templates/hcl_templates/onlyexcept

stevenhorsman reviewed (commented):
I think I've run the documentation through correctly and got an nginx pod running and just have a few minor comments.

One thing that concerns me though, was that the qcow2 file that was built wasn't pre-pended with `se`:
```
Downloading file from image se_podvm_s390x:latest
Uploading file podvm-bc64906-dirty-s390x.qcow2
```
which I was expecting, so I'm not sure if I did something wrong and the image isn't an SE image?
Sorry for the basic question, but is there a way to demonstrate that it is running correctly as an SE VSI as I'm not sure if:
```
$ ibmcloud is ins
Listing instances in all resource groups and region ca-tor under account Steven Horsman's Account as user steven@uk.ibm.com...
ID                                          Name                                Status    Reserved IP   Floating IP     Profile    Image                                VPC                              Zone       Resource group
02r7_23f5d5d7-70fb-40f8-98e3-98f354dab211   podvm-nginx-f3c03534                running   10.249.64.7   -               bz2e-2x8   podvm-bc64906-dirty-s390x            sh-sm-z-cluster-2023-05-09-vpc   ca-tor-2   default
```
is enough? and if so maybe we can add that to the docs?

liudalibj commented:
> I think I've run the documentation through correctly and got an nginx pod running and just have a few minor comments.
> 
> One thing that concerns me though, was that the qcow2 file that was built wasn't pre-pended with `se`:
> 
> ```
> Downloading file from image se_podvm_s390x:latest
> Uploading file podvm-bc64906-dirty-s390x.qcow2
> ```
> 
> which I was expecting, so I'm not sure if I did something wrong and the image isn't an SE image? Sorry for the basic question, but is there a way to demonstrate that it is running correctly as an SE VSI as I'm not sure if:
> 
> ```
> $ ibmcloud is ins
> Listing instances in all resource groups and region ca-tor under account Steven Horsman's Account as user steven@uk.ibm.com...
> ID                                          Name                                Status    Reserved IP   Floating IP     Profile    Image                                VPC                              Zone       Resource group
> 02r7_23f5d5d7-70fb-40f8-98e3-98f354dab211   podvm-nginx-f3c03534                running   10.249.64.7   -               bz2e-2x8   podvm-bc64906-dirty-s390x            sh-sm-z-cluster-2023-05-09-vpc   ca-tor-2   default
> ```
> 
> is enough? and if so maybe we can add that to the docs?

@stevenhorsman thanks for your comments.  To check a vsi is se-enabled, you can check the profile name contain `e` char, and I am trying to update one e2e test case to verify whether the podvm instance is se enabled.

liudalibj commented:
@stevenhorsman 
how to verify the created VSI is SE enabled from https://www.ibm.com/docs/en/linux-on-systems?topic=tasks-enable-kvm-host

inside the vsi:
```
cat /sys/firmware/uv/prot_virt_guest
```
it will return `1` when we are in IBM Secure Execution mode.

Run the same command inside a no-SE vsi.
```
root@no-se-vsi:~# cat /sys/firmware/uv/prot_virt_guest
cat: /sys/firmware/uv/prot_virt_guest: No such file or directory
root@no-se-vsi:~#
```

More information from https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/9/html/configuring_and_managing_virtualization/securing-virtual-machines-in-rhel_configuring-and-managing-virtualization#setting-up-secure-execution-on-ibm-z_securing-virtual-machines-in-rhel
```
peerpod@se-test-39:~$ grep facilities /proc/cpuinfo | grep 158
facilities      : 0 1 2 3 4 6 7 8 9 13 16 17 18 19 20 21 22 23 24 25 26 27 28 30 31 32 33 34 35 36 37 38 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 57 58 59 61 64 73 74 75 76 77 80 81 128 129 130 131 133 134 135 138 139 141 146 148 150 151 152 155 156 158 169
peerpod@se-test-39:~$ ls /sys/firmware | grep uv
uv
```
from the no-se vsi
```
root@no-se-vsi:~# grep facilities /proc/cpuinfo | grep 158
root@no-se-vsi:~# ls /sys/firmware | grep uv
```

liudalibj commented:
I comment out https://github.com/liudalibj/cloud-api-adaptor/blob/se-image-packer/podvm/qcow2/misc-settings.sh#L15 in my local to create the test SE-qcow image which can let me login with `peerpod/PeerP0d`

liudalibj commented on podvm/qcow2/build-se-image-post.sh:
```diff
@@ -0,0 +1,12 @@
+if [ "${SE_BOOT:-0}" != "1" ]; then
+    rm -rf "se-${IMAGE_NAME}"
```
done

liudalibj commented on podvm/qcow2/build-se-image.sh:
```diff
@@ -0,0 +1,163 @@
+    host_keys+="-k ${i} "
+done
+[[ -z $host_keys ]] && echo "Didn't find host key files, please download host key files to `files` folder " && exit 1
+export host_keys
```
done

liudalibj commented on podvm/qcow2/build-se-image.sh:
```diff
@@ -0,0 +1,163 @@
+export host_keys
+export workdir=$(pwd)
+export disksize=100G
+export device=$(lsblk |grep 100G | awk '{ print $1 }')
```
done

liudalibj commented on ibmcloud/SECURE_EXECUTION.md:
```diff
@@ -1,37 +1,38 @@
 
-1. **Build genprotimg on worker node**
+#### Build the Secure Execution enabled peer pod VM image
+You can do this by following the process [documented](../podvm/README.md). Simply you need follow steps which run under `cloud-api-adaptor/podvm` folder:
```
done

liudalibj commented on ibmcloud/SECURE_EXECUTION.md:
````diff
@@ -1,37 +1,38 @@
+The host key must match the host system for which the podvm image is prepared. You can download a host key document from Resource Link.
 
 As a registered user, access the search page:
 ```
````
done

liudalibj commented on ibmcloud/SECURE_EXECUTION.md:
````diff
@@ -42,109 +43,84 @@ If you have never signed in to Resource Link, you need to register before you ca
+> ├── etc
+> └── usr
 > ```
 > - Get the **machine type** and **machine serial number** from worker node only works when there is only one s390x arch host system in the target zone, if there are multiple s390x arch host systems in the target zone, please try to create z vsis on the target s390x arch host system one by one then get the **machine type** and **machine serial number** from the created z vsis on diferent host sys ...
 > - Another way maybe you can contact the system admin to get the **machine type** and **machine serial number** for target host systems
````
Agree, we should figure out a better way to get host key documents, here is a related discuss in internal slack https://ibm-systems-z.slack.com/archives/CA8B73TQE/p1681460192824699

liudalibj commented on ibmcloud/SECURE_EXECUTION.md:
```diff
@@ -42,109 +43,84 @@ If you have never signed in to Resource Link, you need to register before you ca
 
-### Troubleshooting
-The build se-image will use `/dev/nbd0` and `/dev/nbd1`, please make sure they are avaliable to use
+The sample command, assume your working dir is `~/cloud-api-adaptor/ibmcloud/image`:
```
done

stevenhorsman commented:
> @stevenhorsman
how to verify the created VSI is SE enabled from https://www.ibm.com/docs/en/linux-on-systems?topic=tasks-enable-kvm-host

Thanks very much for this info. I think it might be worth linking to in the doc?

Unfortunately my suspicion that my pod wasn't SE, despite having the `bz2e-2x8` type was true:

```
$ ibmcloud is ins
Listing instances in all resource groups and region ca-tor under account Steven Horsman's Account as user steven@uk.ibm.com...
ID                                          Name                                Status    Reserved IP   Floating IP     Profile    Image                                VPC                              Zone       Resource group
02r7_23f5d5d7-70fb-40f8-98e3-98f354dab211   podvm-nginx-f3c03534                running   10.249.64.7   163.74.86.107   bz2e-2x8   podvm-bc64906-dirty-s390x            sh-sm-z-cluster-2023-05-09-vpc   ca-tor-2   default
stevenh@stevenh-sm-se1:~$ ssh root@163.74.86.107
...
root@podvm-nginx-f3c03534:~# cat /sys/firmware/uv/prot_virt_guest
cat: /sys/firmware/uv/prot_virt_guest: No such file or directory
root@podvm-nginx-f3c03534:~# grep facilities /proc/cpuinfo | grep 158
root@podvm-nginx-f3c03534:~# ls /sys/firmware | grep uv
root@podvm-nginx-f3c03534:~#
```

So I'll re-do the peer pod creation to try and work out what I did wrong.

liudalibj commented:
@yoheiueda @bpradipt @stevenhorsman Thanks your review comments, I updated the codes base on them, would you like help to review again when you got a time, thanks!

liudalibj commented:
@stevenhorsman please try to build the se-image with
```
docker build -t se_podvm_s390x \
         --build-arg ARCH=s390x \
         --build-arg CAA_SRC=https://github.com/liudalibj/cloud-api-adaptor \
         --build-arg CAA_SRC_REF=se-image-packer \
         --build-arg SE_BOOT=1 \
         --build-arg BUILDER_IMG=podvm_builder \
         --build-arg BINARIES_IMG=podvm_binaries_s390x \
         --build-arg UBUNTU_IMAGE_URL="" \
         --build-arg UBUNTU_IMAGE_CHECKSUM="" \
         -f Dockerfile.podvm .
```

stevenhorsman commented:
> @stevenhorsman please try to build the se-image with
> 
> ```
> docker build -t se_podvm_s390x \
>          --build-arg ARCH=s390x \
>          --build-arg CAA_SRC=https://github.com/liudalibj/cloud-api-adaptor \
>          --build-arg CAA_SRC_REF=se-image-packer \
>          --build-arg SE_BOOT=1 \
>          --build-arg BUILDER_IMG=podvm_builder \
>          --build-arg BINARIES_IMG=podvm_binaries_s390x \
>          --build-arg UBUNTU_IMAGE_URL="" \
>          --build-arg UBUNTU_IMAGE_CHECKSUM="" \
>          -f Dockerfile.podvm .
> ```

Yes, thanks - I just realised that I followed the doc, so wouldn't have you merged changes in, so am re-running with 
```
--build-arg CAA_SRC=https://github.com/liudalibj/cloud-api-adaptor \
         --build-arg CAA_SRC_REF=se-image-packer \
```
now. Sorry to waste your time!

bpradipt commented on .github/workflows/podvm_builder.yaml:
```diff
@@ -61,6 +61,6 @@ jobs:
           "RUST_VERSION=1.66.0"
           "CAA_SRC=https://github.com/confidential-containers/cloud-api-adaptor"
-          "CAA_SRC_BRANCH=staging"
+          "CAA_SRC_REF=staging"
```
Looks like we are using a mix of CAA_SRC_BRANCH and CAA_SRC_REF which is a mess :-( 

We should make it consistent and uniform across the different Dockerfiles, workflows and doc as part of a separate PR.

@stevenhorsman any preference on whether to use CAA_SRC_BRANCH  or CAA_SRC_REF ?

stevenhorsman replied:
I think that `CAA_SRC_BRANCH` was the original solution and then @wainersm introduced `CAA_SRC_REF` (maybe to other docker files?) for helping to run them against PRs. It's worth getting Wainer's confirmation, but I guess that using (and documenting ) `REF` consistently 'covers' more cases? Not a strong feeling though.

stevenhorsman commented:
> > @stevenhorsman please try to build the se-image with
> > ```
> > docker build -t se_podvm_s390x \
> >          --build-arg ARCH=s390x \
> >          --build-arg CAA_SRC=https://github.com/liudalibj/cloud-api-adaptor \
> >          --build-arg CAA_SRC_REF=se-image-packer \
> >          --build-arg SE_BOOT=1 \
> >          --build-arg BUILDER_IMG=podvm_builder \
> >          --build-arg BINARIES_IMG=podvm_binaries_s390x \
> >          --build-arg UBUNTU_IMAGE_URL="" \
> >          --build-arg UBUNTU_IMAGE_CHECKSUM="" \
> >          -f Dockerfile.podvm .
> > ```
> 
> Yes, thanks - I just realised that I followed the doc, so wouldn't have you merged changes in, so am re-running with
> 
> ```
> --build-arg CAA_SRC=https://github.com/liudalibj/cloud-api-adaptor \
>          --build-arg CAA_SRC_REF=se-image-packer \
> ```
> 
> now. Sorry to waste your time!

With the correct branch/ref used the SE peer pod VM was built correctly and when I ssh into the peer pod VM I see the correct results:

```
root@podvm-nginx-8d700109:~# cat /sys/firmware/uv/prot_virt_guest
1
root@podvm-nginx-8d700109:~# grep facilities /proc/cpuinfo | grep 158
facilities      : 0 1 2 3 4 6 7 8 9 13 16 17 18 19 20 21 22 23 24 25 26 27 28 30 31 32 33 34 35 36 37 38 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 57 58 59 61 64 73 74 75 76 77 80 81 128 129 130 131 133 134 135 138 139 141 146 148 150 151 152 155 156 158 169
root@podvm-nginx-8d700109:~# ls /sys/firmware | grep uv
uv
```
😄

stevenhorsman reviewed (approved):
From a documentation and basic running point of view I've managed to get this running and tested. There might be some minor doc/uniformity improvements we could do, but I don't think they are newly introduced with this PR, so from my (somewhat ignorant) perspective I think this is much better than we had. Thanks for the great work DaLi!

yoheiueda reviewed (approved):
LGTM now. Thank you very much for this great enhancement!

huoqifeng reviewed (approved):
/lgtm, thanks @liudalibj

bpradipt reviewed (approved):
/lgtm
Thanks @liudalibj
